Mathlib/Algebra/Module/LinearMap/Polynomial.lean

```diff
@@ -0,0 +1,336 @@
+that evaluates on an element `x` of `M₁` (represented on the basis `b₁`)
+to the element `f x` of `M₂` (represented on the basis `b₂`). -/
+noncomputable
+def toMvPolynomial (f : M₁ →ₗ[R] M₂) (i : ι₂) :
+    MvPolynomial ι₁ R :=
+  (toMatrix b₁ b₂ f).toMvPolynomial i
+
+lemma toMvPolynomial_eval_eq_apply (f : M₁ →ₗ[R] M₂) (i : ι₂) (c : ι₁ →₀ R) :
+    eval c (f.toMvPolynomial b₁ b₂ i) = b₂.repr (f (b₁.repr.symm c)) i := by
+  rw [toMvPolynomial, Matrix.toMvPolynomial_eval_eq_apply,
+    ← LinearMap.toMatrix_mulVec_repr b₁ b₂, LinearEquiv.apply_symm_apply]
+
+open Algebra.TensorProduct in
+lemma toMvPolynomial_baseChange (f : M₁ →ₗ[R] M₂) (i : ι₂) (A : Type*) [CommRing A] [Algebra R A] :
+    (f.baseChange A).toMvPolynomial (basis A b₁) (basis A b₂) i =
+      MvPolynomial.map (algebraMap R A) (f.toMvPolynomial b₁ b₂ i) := by
+  simp only [toMvPolynomial, toMatrix_baseChange, Matrix.toMvPolynomial_map]
+
+lemma toMvPolynomial_isHomogeneous (f : M₁ →ₗ[R] M₂) (i : ι₂) :
+    (f.toMvPolynomial b₁ b₂ i).IsHomogeneous 1 :=
+  Matrix.toMvPolynomial_isHomogeneous _ _
+
+lemma toMvPolynomial_totalDegree_le (f : M₁ →ₗ[R] M₂) (i : ι₂) :
+    (f.toMvPolynomial b₁ b₂ i).totalDegree ≤ 1 :=
+  Matrix.toMvPolynomial_totalDegree_le _ _
+
+@[simp]
+lemma toMvPolynomial_constantCoeff (f : M₁ →ₗ[R] M₂) (i : ι₂) :
+    constantCoeff (f.toMvPolynomial b₁ b₂ i) = 0 :=
+  Matrix.toMvPolynomial_constantCoeff _ _
+
+@[simp]
+lemma toMvPolynomial_zero (i : ι₂) : (0 : M₁ →ₗ[R] M₂).toMvPolynomial b₁ b₂ i = 0 := by
+  simp only [toMvPolynomial, map_zero, Matrix.toMvPolynomial_zero]
+
+lemma toMvPolynomial_add (f g : M₁ →ₗ[R] M₂) (i : ι₂) :
+    (f + g).toMvPolynomial b₁ b₂ i = f.toMvPolynomial b₁ b₂ i + g.toMvPolynomial b₁ b₂ i := by
+  simp only [toMvPolynomial, map_add, Matrix.toMvPolynomial_add]
+
+lemma toMvPolynomial_comp (g : M₂ →ₗ[R] M₃) (f : M₁ →ₗ[R] M₂) (i : ι₃) :
+    (g ∘ₗ f).toMvPolynomial b₁ b₃ i =
+      bind₁ (f.toMvPolynomial b₁ b₂) (g.toMvPolynomial b₂ b₃ i) := by
+  simp only [toMvPolynomial, toMatrix_comp b₁ b₂ b₃, Matrix.toMvPolynomial_mul]
+  rfl
+
+end LinearMap
+
+variable {R L M n ι ιM : Type*}
+
+section LinearAlgebra
+
+variable [CommRing R] [AddCommGroup L] [Module R L] [AddCommGroup M] [Module R M]
+variable (φ : L →ₗ[R] Module.End R M)
+variable [Fintype ι] [Fintype ιM] [DecidableEq ι]
+
+namespace LinearMap
+
+section basic
+
+variable [DecidableEq ιM] (b : Basis ι R L) (bₘ : Basis ιM R M)
+
+open Matrix
+
+/-- Let `L` and `M` be modules over `R`, and let `φ : L →ₗ[R] Module.End R M` be a linear map.
+Let `b` be a basis of `L` and `bₘ` a basis of `M`.
+Then `LinearMap.polyCharpoly φ b bₘ` is the polynomial that evaluates on elements `x` of `L`
+to the characteristic polynomial of `φ x` acting on `M`.
+
+This definition does not depend on the choice of `bₘ`
+(see `LinearMap.polyCharpoly_basisIndep`). -/
+noncomputable
+def polyCharpoly : Polynomial (MvPolynomial ι R) :=
```

Comment:
I wonder if it is useful to package `polyCharpoly φ b (Module.Free.chooseBasis R M)` into a new function so that the user won't need to obtain a basis before using this?

Reply:
I agree with @erdOne . I would rename this to `polyCharpolyAux` and then once `polyCharpolyAux_basisIndep` has been proved, define a new `polyCharpoly` that invokes `Module.Free.chooseBasis R M`.

Reply:
Done